### PR TITLE
Include full OrcType in StreamDescriptor

### PR DESCRIPTION
**Pull request by bhhari**
SelectiveStreamReaders for varchars and decimals need additional type information such as precision, scale and length. 
 
See #13306

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-orc/src/main/java/com/facebook/presto/orc/StreamDescriptor.java [hunk 1]
```diff
@@ -27,23 +28,23 @@
...
 
-    public StreamDescriptor(String streamName, int streamId, String fieldName, OrcTypeKind streamType, OrcDataSource orcDataSource, List<StreamDescriptor> nestedStreams, int sequence)
+    public StreamDescriptor(String streamName, int streamId, String fieldName, OrcType orcType, OrcDataSource orcDataSource, List<StreamDescriptor> nestedStreams, int sequence)
     {
         this.streamName = requireNonNull(streamName, "streamName is null");
         this.streamId = streamId;
         this.sequence = sequence;
         this.fieldName = requireNonNull(fieldName, "fieldName is null");
-        this.streamType = requireNonNull(streamType, "type is null");
+        this.orcType = requireNonNull(orcType, "type is null");
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/StreamDescriptor.java]: nit: "type is null" -> "orcType is null"

File: presto-orc/src/main/java/com/facebook/presto/orc/metadata/OrcType.java [hunk 2]
```diff
@@ -282,4 +282,9 @@ public String toString()
 
         return orcTypes;
     }
+
+    public static OrcType createOrcType(OrcTypeKind orcTypeKind)
+    {
+        return new OrcType(orcTypeKind);
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/metadata/OrcType.java]: OrcTypeKind is not enough to create a valid instance of OrcType. This method is only used in a test, hence, I suggest to remove it and update the caller like so:

```
    private static StreamDescriptor makeStreamDescriptor(int levels)
    {
        DummyOrcDataSource orcDataSource = new DummyOrcDataSource();

        OrcType intType = new OrcType(INT, ImmutableList.of(), ImmutableList.of(), Optional.empty(), Optional.empty(), Optional.empty());
        OrcType listType = new OrcType(LIST, ImmutableList.of(1), ImmutableList.of("item"), Optional.empty(), Optional.empty(), Optional.empty());

        StreamDescriptor streamDescriptor = new StreamDescriptor("a", 0, "a", intType, orcDataSource, ImmutableList.of());
        for (int i = 0; i < levels; i++) {
            streamDescriptor = new StreamDescriptor("a", 0, "a", listType, orcDataSource, ImmutableList.of(streamDescriptor));
        }

        return streamDescriptor;
    }
```

File: presto-orc/src/main/java/com/facebook/presto/orc/StreamDescriptor.java [hunk 3]
```diff
@@ -65,7 +66,12 @@ public int getSequence()
 
     public OrcTypeKind getStreamType()
```
Comment (mbasmanova) [inline presto-orc/src/main/java/com/facebook/presto/orc/StreamDescriptor.java]: getStreamType name doesn't work well with getOrcType. Would you rename getStreamType to getOrcTypeKind in a separate commit?

**Comment by oerling** [comment]
This seems to be an appropriate solution for a few upcoming issues:

*	When reading structs/lists./maps for filtering only, the full type information corresponding to the file will be needed.
*	Making coercion for values that are extracted.
*	Dealing with varchar lengths and char padding.

 

This is needed for Slice readers for the truncation behavior. Note that if this type and the type that is requested from the plan differ, then we would get different results.

 

Consider the case of coercion in a struct: You have a physical int that is coerced to a logical short. You have a filter < 100000. This will be false for value 90000 because this coerces to null, not being castable to short. This case seems to indicate that the type of a column as it is represented in the plan (PageSource) still needs to be passed down all the way through the reader tree, also in cases where reading is for filter only. 

 

So, maybe we pass the type requested by the plan (the type in the PageSource) all the way down the reader tree also for cases that are  filter only. So outputType would not be optional, rather there would be a separate flag for whether the value is materialized.

 

Masha, do you see another way for dealing with this? Do you think we should deal with this now or should we just get the string readers functional without the type passing. For the latter purpose this PR is entirely fine and may turn out handy for future applications too.

 

 

***

 

In getOrcType getter, redundant this in ‘this.orcType’

**Comment by mbasmanova** [comment]
@oerling 

> So, maybe we pass the type requested by the plan (the type in the PageSource) all the way down the reader tree also for cases that are  filter only. So outputType would not be optional, rather there would be a separate flag for whether the value is materialized.

Yes. I do believe this will be necessary once we get to deal with schema evolution and coercion. At the same time I think the change in this PR is also useful. That change is sufficient for right now, hence, I'd go with it.

**Comment by mbasmanova** [comment]
@bhhari Looks good to me % we still need a first commit that renames getStreamType to getOrcTypeKind.

**Comment by bhhari** [comment]
> @bhhari Looks good to me % we still need a first commit that renames getStreamType to getOrcTypeKind.

I was planning on that commit after this.

**Comment by oerling** [comment]
Yes, let’s go with this.

**Pull request merged**